### PR TITLE
Initial commit for DNIS pooling sample

### PR DESCRIPTION
**Pull request by goldenwitch**
Still needs a better readme explaining DNIS pooling, I didn't get the chance to write that this week.
I did clean up the library bits a tiny amount, so this should be ready to go in as an experimental option.

### CODE REVIEW

File: samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/Controllers/BrokerController.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Microsoft.BotBuilderSamples.Controllers
+{
+    [ApiController]
+    [Route("broker")]
+    public class BrokerController : ControllerBase
+    {
+        private readonly IDictionary<string, JObject> _storage;
```
Comment (arturl) [inline samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/Controllers/BrokerController.cs:17]: Should this be durable storage?

File: samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                            pool[pair.Key] = true;
+                            return pair.Key;
+                        }
+                    }
+                }
+                finally
+                {
+                    latchingSemaphore.Release();
+                }
+                await Task.Delay(10); //We need the task.delay in case this thread is the one blocking the release.
```
Comment (arturl) [inline samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs:52]: huh?

File: samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        /// Finds the next unlatched key, latches it and returns it. Must be thread safe.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<string> RequestKey()
+        {
+            while (true)
+            {
+                try
+                {
+                    await latchingSemaphore.WaitAsync();
```
Comment (arturl) [inline samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs:36]: You could use a regular lock and a regular dictionary here, then this method doesn't need to be async (short-term locks on data access are fine, since there is no I/O)

File: samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace Microsoft.BotBuilderSamples
+{
+    /// <summary>
+    /// Holds list of keys with latching condition
+    /// </summary>
+    public class LatchingKeyPool
+    {
+        private SemaphoreSlim latchingSemaphore = new SemaphoreSlim(1, 1);
+        private readonly ConcurrentDictionary<string, bool> pool = new ConcurrentDictionary<string, bool>(StringComparer.InvariantCultureIgnoreCase);
```
Comment (arturl) [inline samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs:16]: Durable storage here too?

File: samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs (hunk 2 above)
Comment (goldenwitch) [reply samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs]: I had a good reason for this one at the time, but I can't remember the details to the edge case, and my tests don't capture said edge case so I got rid of it :D

File: samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs (hunk 4 above)
Comment (goldenwitch) [reply samples/csharp_dotnetcore/07.telephony-pooling-contextual-transfer/LatchingKeyPool.cs]: Yeah making a version that supports local and durable storage would be WAY better. I feel like this is just an example implementation for in memory storage, but we should probably author an Azure friendly durable storage version of this. I pulled out the interface so that's easier to do now.

**Review by arturl — APPROVED** [review]
Ship it!
